### PR TITLE
Expose method to warm connection pool from JedisPool #1828

### PR DESCRIPTION
**Pull request by tmack8001**
I decided to expose a public method on the `AbstractJedisPool` to allow users to then explicitly warm their `JedisPool` after creation instead of adding yet another parameter to the already overwritten set of constructors for the `JedisPool`. This method wraps the call to the internal pool catching exceptions and logging the underlying error that occurred.

Example usage would be the following:
```
JedisPool jedisPool = new JedisPool(new URI("localhost:6380"));
jedisPool.preparePool(); // explicit warming of the internal connection pool for Jedis
```

The logic within preparePool is standard within the `GenericObjectPool` implementation to make sure that there is always a minimum of `minIdle` connections available within the pool. This helps in high throughput applications as then the cost of creating connections are done before hand vs upon demand/use.

### CODE REVIEW

**Comment by tmack8001** [comment]
Not sure of the failure, but will try to investigate why the following test failed.

```
Failed tests: 
  JedisSentinelPoolTest.ensureSafeTwiceFailover:95->forceFailover:190 expected:<[foobared]> but was:<[1b58ee375b42e41f0e48ef2ff27d10a5b1f6924a9acdcdba7cae868e7adce6bf]>
```

File: src/main/java/redis/clients/jedis/JedisPool.java [hunk 1]
```diff
@@ -30,11 +30,11 @@ public JedisPool(final String host) {
     URI uri = URI.create(host);
     if (JedisURIHelper.isValid(uri)) {
       this.internalPool = new GenericObjectPool<Jedis>(new JedisFactory(uri,
-          Protocol.DEFAULT_TIMEOUT, Protocol.DEFAULT_TIMEOUT, null), new GenericObjectPoolConfig());
+          Protocol.DEFAULT_TIMEOUT, Protocol.DEFAULT_TIMEOUT, null), new GenericObjectPoolConfig<Jedis>());
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisPool.java]: Do these changes necessary?

File: src/main/java/redis/clients/jedis/JedisPoolAbstract.java [hunk 2]
```diff
@@ -3,10 +3,14 @@
 import org.apache.commons.pool2.PooledObjectFactory;
 import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisPoolAbstract.java]: Separate these imports from `redis.*` package.

File: src/main/java/redis/clients/jedis/JedisPoolAbstract.java [hunk 3]
```diff
@@ -15,6 +19,14 @@ public JedisPoolAbstract(GenericObjectPoolConfig poolConfig, PooledObjectFactory
     super(poolConfig, factory);
   }
 
+  public void prepareInternalPool() {
+    try {
+      this.internalPool.preparePool();
+    } catch (Exception e) {
+      log.warn("exception occurred warming jedis internal connection pool", e);
+    }
+  }
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisPoolAbstract.java]: Put this method in `Pool` class.

File: src/main/java/redis/clients/jedis/JedisPoolAbstract.java [hunk 4]
```diff
@@ -3,10 +3,14 @@
 import org.apache.commons.pool2.PooledObjectFactory;
 import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import redis.clients.jedis.util.Pool;
 
 public class JedisPoolAbstract extends Pool<Jedis> {
 
+  protected Logger log = LoggerFactory.getLogger(getClass().getName());
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisPoolAbstract.java]:
```suggestion
  private Logger logger = LoggerFactory.getLogger(getClass().getName());
```

File: src/main/java/redis/clients/jedis/JedisPoolAbstract.java [hunk 5]
```diff
@@ -15,6 +19,14 @@ public JedisPoolAbstract(GenericObjectPoolConfig poolConfig, PooledObjectFactory
     super(poolConfig, factory);
   }
 
+  public void prepareInternalPool() {
+    try {
+      this.internalPool.preparePool();
+    } catch (Exception e) {
+      log.warn("exception occurred warming jedis internal connection pool", e);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisPoolAbstract.java]: More *sound* message, please.

File: src/test/java/redis/clients/jedis/tests/JedisPoolTest.java [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertThat;
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/tests/JedisPoolTest.java]: No need.
```suggestion
```

File: src/main/java/redis/clients/jedis/JedisPool.java (hunk 1 above)
Comment (tmack8001) [reply src/main/java/redis/clients/jedis/JedisPool.java]: no, but does resolve a compiler warning about unchecked assignment of a parameterized class

Comment (tmack8001) [reply src/main/java/redis/clients/jedis/JedisPool.java]: though you do bring up a point that there are other constructors that should have this added to them as well to really clean up these compiler warnings, but are unrelated to this specific change

File: src/main/java/redis/clients/jedis/util/Pool.java [hunk 7]
```diff
@@ -45,6 +49,14 @@ public void initPool(final GenericObjectPoolConfig poolConfig, PooledObjectFacto
     this.internalPool = new GenericObjectPool<T>(factory, poolConfig);
   }
 
+  public void prepareInternalPool() {
+    try {
+      this.internalPool.preparePool();
+    } catch (Exception e) {
+      logger.warn("failed to initialize minIdle connections in the internal jedis connection pool", e);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/util/Pool.java]: Please change:
- Capitalize beginning of sentence.
- Remove `jedis`. This is a generic class (`Pool<T>`) and T may or may not be Jedis.

```suggestion
      logger.warn("Could not initialize minimum number of idle resources", e);
```

File: src/test/java/redis/clients/jedis/tests/JedisPoolTest.java [hunk 8]
```diff
@@ -401,6 +400,36 @@ public void testCloseConnectionOnMakeObject() {
...
+  }
+
+  @Test
+  public void testPreparePoolCatchesExceptions() {
+    JedisPoolConfig jedisPoolConfig = new JedisPoolConfig();
+    jedisPoolConfig.setMinIdle(8);
+
+    JedisPool pool = new JedisPool(jedisPoolConfig, "invalid", hnp.getPort(), 2000);
+    assertEquals(pool.getNumIdle(), 0);
+    pool.prepareInternalPool(); // catches and logs exception
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/tests/JedisPoolTest.java]: Add a TODO comment to mock the logger and assert the log.

Comment (sazzad16) [reply src/test/java/redis/clients/jedis/tests/JedisPoolTest.java]:
```suggestion
    pool.prepareInternalPool(); // catches and logs exception
    // TODO: mock logger and check logs
```

File: src/main/java/redis/clients/jedis/JedisPoolAbstract.java [hunk 9]
```diff
@@ -2,6 +2,8 @@
 
 import org.apache.commons.pool2.PooledObjectFactory;
 import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisPoolAbstract.java]:
```suggestion
```

File: src/main/java/redis/clients/jedis/util/Pool.java [hunk 10]
```diff
@@ -7,11 +7,16 @@
...
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import redis.clients.jedis.exceptions.JedisConnectionException;
 import redis.clients.jedis.exceptions.JedisException;
 import redis.clients.jedis.exceptions.JedisExhaustedPoolException;
 
 public abstract class Pool<T> implements Closeable {
+  private Logger logger = LoggerFactory.getLogger(getClass().getName());
```
Comment (gkorland) [inline src/main/java/redis/clients/jedis/util/Pool.java]: `private final`

Comment (sazzad16) [inline src/main/java/redis/clients/jedis/util/Pool.java]:
```suggestion
  private final Logger logger = LoggerFactory.getLogger(getClass().getName());
```

**Review by sazzad16 — DISMISSED** [review]

File: src/main/java/redis/clients/jedis/util/Pool.java [hunk 11]
```diff
@@ -45,6 +49,14 @@ public void initPool(final GenericObjectPoolConfig poolConfig, PooledObjectFacto
     this.internalPool = new GenericObjectPool<>(factory, poolConfig);
   }
 
+  public void prepareInternalPool() {
+    try {
+      this.internalPool.preparePool();
+    } catch (Exception e) {
```
Comment (gkorland) [inline src/main/java/redis/clients/jedis/util/Pool.java:55]: Why are we hiding the exception and not throwing it back to the client?

**Pull request closed without merging**